File: src/Core/LotMatching/Services/LotMatcher.py
import copy
from dataclasses import dataclass
from typing import Self, Sequence, TypeVar

from arrow import Arrow

from Core.FinancialEvents.Schemas.CommonFormats import GenericShortLong
from Core.FinancialEvents.Schemas.Events import TradeEvent
from Core.FinancialEvents.Schemas.Lots import TaxLot
from Core.LotMatching.Contracts.LotMatchingMethod import LotMatchingMethod
from Core.LotMatching.Schemas.Lot import Lot
from Core.LotMatching.Schemas.Trade import Trade

TAX_LOT_TYPE = TypeVar("TAX_LOT_TYPE", bound=TaxLot[TradeEvent, TradeEvent], covariant=True)


@dataclass
class LotMatchingDetails:
    Lots: Sequence[Lot]
    Trades: Sequence[Trade]


@dataclass
class GeneratedLotWithTradeEvents:
    Lot: Lot
    AcquiredTrade: TradeEvent
    SoldTrade: TradeEvent


class GenericLotMatchingDetails:
    Lots: Sequence[TaxLot[TradeEvent, TradeEvent]]
    Trades: Sequence[TradeEvent]

    def __init__(self, lots: Sequence[TaxLot[TradeEvent, TradeEvent]], trades: Sequence[TradeEvent]):
        self.Lots = lots
        self.Trades = trades

    def getTradesOfLotsClosedInPeriod(self, periodStart: Arrow, periodEnd: Arrow) -> Self:
        lotsClosedInPeriod = list(filter(lambda lot: lot.Sold.Date >= periodStart and lot.Sold.Date <= periodEnd, self.Lots))

        lotAcquiredTrades = list(map(lambda lot: lot.Acquired, lotsClosedInPeriod))
        lotSoldTrades = list(map(lambda lot: lot.Sold, lotsClosedInPeriod))
        allTrades = lotAcquiredTrades + lotSoldTrades

        sortedTrades = sorted(allTrades, key=lambda trade: trade.Date)

        return self.__class__(lots=lotsClosedInPeriod, trades=sortedTrades)


class LotMatcher:

    def matchLotsWithTrades(self, method: LotMatchingMethod, events: Sequence[Trade]) -> LotMatchingDetails:
        lots = method.performMatching(events)

        trades = method.generateTradesFromLotsWithTracking(lots)

        details = LotMatchingDetails(Lots=lots, Trades=trades)

        return details

    def matchLotsWithGenericTradeEvents(self, method: LotMatchingMethod, events: Sequence[TradeEvent]) -> GenericLotMatchingDetails:

        tradeEventMappings: dict[str, TradeEvent] = dict()
        tradeMappings: dict[str, Trade] = dict()
        convertedGeneratedMappings: dict[str, TradeEvent] = dict()

        for event in events:
            convertedEvent = self._convertTradeEvent(event)
            tradeEventMappings[event.ID] = event
            tradeMappings[convertedEvent.ID] = convertedEvent

        convertedEvents = list(tradeMappings.values())
        matchingDetails = self.matchLotsWithTrades(method=method, events=convertedEvents)
        generatedLots = matchingDetails.Lots

        # this part seems dangerous
        generatedLotsWithTradeEvents = list(map(lambda lot: self._generateTradeEventBasedOnLot(lot, tradeEventMappings), generatedLots))

        # TODO: Get rid of ShortLong type on GenericTaxLot, this is going to have to live on some other layer (or there is a missing layer between the entities here)
        def convertLot(lot: GeneratedLotWithTradeEvents) -> TaxLot[TradeEvent, TradeEvent]:
            lotId = lot.Lot.Acquired.Relation.ID
            acquiredTrade = lot.AcquiredTrade
            soldTrade = lot.SoldTrade

            newLot = TaxLot(
                ID=lotId,
                ISIN=acquiredTrade.ISIN,
                Quantity=lot.Lot.Quantity,
                Acquired=acquiredTrade,
                Sold=soldTrade,
                ShortLongType=GenericShortLong.LONG,  # TODO: we can check trade dates? if it was sold before it was bought, it's a short
            )

            return newLot

        convertedLots = list(map(convertLot, generatedLotsWithTradeEvents))

        # for clarity, list all buys and then all sells for trades
        buys = list(map(lambda lot: lot.AcquiredTrade, generatedLotsWithTradeEvents))
        sells = list(map(lambda lot: lot.SoldTrade, generatedLotsWithTradeEvents))
        allTradesTakenFromLots = buys + sells

        details = GenericLotMatchingDetails(lots=convertedLots, trades=allTradesTakenFromLots)

        return details

    def _convertTradeEvent(self, event: TradeEvent) -> Trade:
        trade = Trade(ID=event.ID, Quantity=event.ExchangedMoney.UnderlyingQuantity, Date=event.Date)
        return trade

    def _generateTradeEventBasedOnLot(self, lot: Lot, tradeEventMappings: dict[str, TradeEvent]) -> GeneratedLotWithTradeEvents:
        acquiredTrade = lot.Acquired
        soldTrade = lot.Sold

        underlyingAcquiredTrade = tradeEventMappings.get(acquiredTrade.Relation.ID)
        underlyingSoldTrade = tradeEventMappings.get(soldTrade.Relation.ID)

        if underlyingAcquiredTrade is None or underlyingSoldTrade is None:
            raise ValueError("Acquired Trade or Sold Trade is missing lookup")

        lotQuantity = lot.Quantity
        clonedAcquiredTrade = copy.deepcopy(underlyingAcquiredTrade)
        clonedAcquiredTrade.ExchangedMoney.UnderlyingQuantity = lotQuantity

        clonedSoldTrade = copy.deepcopy(underlyingSoldTrade)
        clonedSoldTrade.ExchangedMoney.UnderlyingQuantity = -lotQuantity

        generatedLotWithTradeEvents = GeneratedLotWithTradeEvents(Lot=lot, AcquiredTrade=clonedAcquiredTrade, SoldTrade=clonedSoldTrade)

        return generatedLotWithTradeEvents

File: lib/Core/LotMatching/LotMatcher.ts
import type { DateTime } from "luxon";
import { GenericShortLong } from "@brrr/Core/Schemas/CommonFormats.ts";
import type { TradeEvent } from "@brrr/Core/Schemas/Events.ts";
import type { TaxLot } from "@brrr/Core/Schemas/Lots.ts";
import type { Lot } from "./Lot.ts";
import type { LotMatchingMethod } from "./LotMatchingMethod.ts";
import type { Trade } from "./Trade.ts";

export type LotMatchingDetails = {
	lots: Lot[];
	trades: Trade[];
};

type GeneratedLotWithTradeEvents = {
	lot: Lot;
	acquiredTrade: TradeEvent;
	soldTrade: TradeEvent;
};

export class GenericLotMatchingDetails {
	lots: TaxLot<TradeEvent, TradeEvent>[];
	trades: TradeEvent[];

	constructor(lots: TaxLot<TradeEvent, TradeEvent>[], trades: TradeEvent[]) {
		this.lots = lots;
		this.trades = trades;
	}

	getTradesOfLotsClosedInPeriod(periodStart: DateTime, periodEnd: DateTime): GenericLotMatchingDetails {
		const lotsClosedInPeriod = this.lots.filter(
			(lot) => lot.sold.date >= periodStart && lot.sold.date <= periodEnd,
		);

		const lotAcquiredTrades = lotsClosedInPeriod.map((lot) => lot.acquired);
		const lotSoldTrades = lotsClosedInPeriod.map((lot) => lot.sold);
		const allTrades = [...lotAcquiredTrades, ...lotSoldTrades];

		const sortedTrades = [...allTrades].sort((a, b) => a.date.toMillis() - b.date.toMillis());

		return new GenericLotMatchingDetails(lotsClosedInPeriod, sortedTrades);
	}
}

export class LotMatcher {
	matchLotsWithTrades(method: LotMatchingMethod, events: Trade[]): LotMatchingDetails {
		const lots = method.performMatching(events);
		const trades = method.generateTradesFromLotsWithTracking(lots);
		return { lots, trades };
	}

	matchLotsWithGenericTradeEvents(method: LotMatchingMethod, events: TradeEvent[]): GenericLotMatchingDetails {
		const tradeEventMappings = new Map<string, TradeEvent>();
		const tradeMappings = new Map<string, Trade>();

		for (const event of events) {
			const convertedEvent = this._convertTradeEvent(event);
			tradeEventMappings.set(event.id, event);
			tradeMappings.set(convertedEvent.id, convertedEvent);
		}

		const convertedEvents = [...tradeMappings.values()];
		const matchingDetails = this.matchLotsWithTrades(method, convertedEvents);
		const generatedLots = matchingDetails.lots;

		const generatedLotsWithTradeEvents = generatedLots.map((lot) => this._generateTradeEventBasedOnLot(lot, tradeEventMappings));

		const convertedLots: TaxLot<TradeEvent, TradeEvent>[] = generatedLotsWithTradeEvents.map((lotWithEvents) => {
			const lotId = lotWithEvents.lot.acquired.relation.id;
			return {
				id: lotId,
				financialIdentifier: lotWithEvents.acquiredTrade.financialIdentifier,
				quantity: lotWithEvents.lot.quantity,
				acquired: lotWithEvents.acquiredTrade,
				sold: lotWithEvents.soldTrade,
				shortLongType: GenericShortLong.LONG,
				provenance: [],
			} as TaxLot<TradeEvent, TradeEvent>;
		});

		const buys = generatedLotsWithTradeEvents.map((l) => l.acquiredTrade);
		const sells = generatedLotsWithTradeEvents.map((l) => l.soldTrade);
		const allTradesTakenFromLots = [...buys, ...sells];

		return new GenericLotMatchingDetails(convertedLots, allTradesTakenFromLots);
	}

	private _convertTradeEvent(event: TradeEvent): Trade {
		return { id: event.id, quantity: event.exchangedMoney.underlyingQuantity, date: event.date };
	}

	private _generateTradeEventBasedOnLot(
		lot: Lot,
		tradeEventMappings: Map<string, TradeEvent>,
	): GeneratedLotWithTradeEvents {
		const acquiredRelation = lot.acquired.relation;
		const soldRelation = lot.sold.relation;

		const underlyingAcquiredTrade = tradeEventMappings.get(acquiredRelation.id);
		const underlyingSoldTrade = tradeEventMappings.get(soldRelation.id);

		if (!underlyingAcquiredTrade || !underlyingSoldTrade) {
			throw new Error("Acquired Trade or Sold Trade is missing lookup");
		}

		const lotQuantity = lot.quantity;

		// Spread copy instead of deepcopy
		const clonedAcquiredTrade: TradeEvent = {
			...underlyingAcquiredTrade,
			exchangedMoney: {
				...underlyingAcquiredTrade.exchangedMoney,
				underlyingQuantity: lotQuantity,
				comissionTotal: (1 / underlyingAcquiredTrade.exchangedMoney.underlyingQuantity) *
					lotQuantity *
					underlyingAcquiredTrade.exchangedMoney.comissionTotal,
			},
		};

		const clonedSoldTrade: TradeEvent = {
			...underlyingSoldTrade,
			exchangedMoney: {
				...underlyingSoldTrade.exchangedMoney,
				underlyingQuantity: -lotQuantity,
				comissionTotal: (1 / underlyingSoldTrade.exchangedMoney.underlyingQuantity) *
					-lotQuantity *
					underlyingSoldTrade.exchangedMoney.comissionTotal,
			},
		};

		return { lot, acquiredTrade: clonedAcquiredTrade, soldTrade: clonedSoldTrade };
	}
}
